File: src/csharp/Intel/Iced/Intel/FastFormatterInternal/FmtData.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if FAST_FMT
using System;
using Iced.Intel.FormatterInternal;
using Iced.Intel.Internal;

namespace Iced.Intel.FastFormatterInternal {
	static partial class FmtData {
		public static readonly string[] Mnemonics = ParseData(out Flags);
		public static readonly FastFmtFlags[] Flags;

		static string[] ParseData(out FastFmtFlags[] outFlags) {
			var reader = new DataReader(GetSerializedData());
			var strings = FormatterStringsTable.GetStringsTable();
			var mnemonics = new string[IcedConstants.CodeEnumCount];
			var flags = new FastFmtFlags[IcedConstants.CodeEnumCount];

			var ca = new char[1];
			int prevIndex = -1;
			var prevFlags = FastFmtFlags.None;
			for (int i = 0; i < mnemonics.Length; i++) {
				var f = (FastFmtFlags)reader.ReadByte();
				int currentIndex;
				if ((f & FastFmtFlags.SameAsPrev) != 0) {
					currentIndex = reader.Index;
					reader.Index = prevIndex;
				}
				else {
					currentIndex = -1;
					prevIndex = reader.Index;
				}
				var mnemonic = strings[reader.ReadCompressedUInt32()];
				if ((prevFlags & FastFmtFlags.HasVPrefix) == (f & FastFmtFlags.HasVPrefix) &&
					(f & FastFmtFlags.SameAsPrev) != 0) {
					mnemonic = mnemonics[i - 1];
				}
				else if ((f & FastFmtFlags.HasVPrefix) != 0) {
					ca[0] = 'v';
					mnemonic = string.Intern(new string(ca) + mnemonic);
				}

				flags[i] = f;
				mnemonics[i] = mnemonic;
				prevFlags = f;
				if (currentIndex >= 0)
					reader.Index = currentIndex;
			}
			if (reader.CanRead)
				throw new InvalidOperationException();

			outFlags = flags;
			return mnemonics;
		}
	}
}
#endif

File: src/java/iced-x86/src/main/java/com/github/icedland/iced/x86/fmt/fast/FmtData.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.fmt.fast;

import com.github.icedland.iced.x86.internal.IcedConstants;
import com.github.icedland.iced.x86.internal.ResourceReader;
import com.github.icedland.iced.x86.internal.fmt.FormatterStringsTable;

@SuppressWarnings("deprecation")
final class FmtData {
	static final String[] mnemonics;
	static final byte[] flags;

	static byte[] getSerializedData() {
		return ResourceReader.readByteArray(FmtData.class.getClassLoader(), "com/github/icedland/iced/x86/fmt/fast/FmtData.bin");
	}

	static {
		com.github.icedland.iced.x86.internal.DataReader reader = new com.github.icedland.iced.x86.internal.DataReader(getSerializedData());
		String[] strings = FormatterStringsTable.getStringsTable();
		String[] mnemonicsTmp = new String[IcedConstants.CODE_ENUM_COUNT];
		byte[] flagsTmp = new byte[IcedConstants.CODE_ENUM_COUNT];

		char[] ca = new char[1];
		int prevIndex = -1;
		byte prevFlags = FastFmtFlags.NONE;
		for (int i = 0; i < mnemonicsTmp.length; i++) {
			byte f = (byte)reader.readByte();
			int currentIndex;
			if ((f & FastFmtFlags.SAME_AS_PREV) != 0) {
				currentIndex = reader.getIndex();
				reader.setIndex(prevIndex);
			}
			else {
				currentIndex = -1;
				prevIndex = reader.getIndex();
			}
			String mnemonic = strings[reader.readCompressedUInt32()];
			if ((prevFlags & FastFmtFlags.HAS_VPREFIX) == (f & FastFmtFlags.HAS_VPREFIX) &&
					(f & FastFmtFlags.SAME_AS_PREV) != 0) {
				mnemonic = mnemonicsTmp[i - 1];
			}
			else if ((f & FastFmtFlags.HAS_VPREFIX) != 0) {
				ca[0] = 'v';
				mnemonic = (new String(ca) + mnemonic).intern();
			}

			flagsTmp[i] = f;
			mnemonicsTmp[i] = mnemonic;
			prevFlags = f;
			if (currentIndex >= 0)
				reader.setIndex(currentIndex);
		}
		if (reader.canRead())
			throw new UnsupportedOperationException();

		flags = flagsTmp;
		mnemonics = mnemonicsTmp;
	}
}
